src/TypesFinder/FindTypeFromAst.php

```diff
@@ -19,10 +19,16 @@ class FindTypeFromAst
      */
     public function __invoke($astType, LocatedSource $locatedSource, $namespace = '')
     {
-        $context = (new ContextFactory())->createForNamespace(
-            $namespace,
-            $locatedSource->getSource()
-        );
+        static $contexts = [];
```

Comment:
Contexts need to be cached both by namespace and the located source (f.e. with a hash).

Just writing this down for the sake of moving on from this today, I suggest creating a new abstract class, f.e. 

```php
class ContextReflection
{
    public function getNamespaceName()
    {
        // ...
    }

    public function getLocatedSource(): LocatedSource
    {
        // ...
    }
}
```

Both `ReflectionClass` and `AbstractReflectionFunction` could then extend this class and the above methods can be removed from them.

A `CachedContextFactory` could then be injected from the Reflector instance.